os/arch/arm/src/amebasmart/amebasmart_serial.c

```diff
@@ -1195,6 +1259,74 @@ static uint32_t rtk_uart_resume(uint32_t expected_idle_time, void *param)
 {
 	(void)expected_idle_time;
 	(void)param;
+
+#if 1
+	ALIGNMTO(CACHE_LINE_SIZE) u8 flag[CACHE_LINE_ALIGMENT(64)];
+	ALIGNMTO(CACHE_LINE_SIZE) u32 uart_data[16];
+
+	/* reset buffer and head pointer for FIFO drain buffer */
+	g_uart1_dataleft = 0;
+	memset(g_uart1_buf, 0, sizeof(g_uart1_buf));
+	g_uart1_buf_head = &g_uart1_buf;
+
+	IPC_MSG_STRUCT ipc_req_msg  __attribute__((aligned(64)));
+	ipc_req_msg.msg_type = IPC_USER_POINT;
+	ipc_req_msg.msg = (u32)uart_data;
+	ipc_req_msg.msg_len = sizeof(uart_data);
+	ipc_req_msg.rsvd = (u32)flag;
+
+	memset(flag, 0, sizeof(flag));
+	memset(uart_data, 0, sizeof(uart_data));
+
+	/* indicate CA32 is ready to rx, switch back to CA32 */
+	uart_data[2] = uart_index_get(g_uart1priv.tx);
+	uart_data[7] = 0; 								// 1 switch to KM4, 0 switch to CA32
+	uart_data[10] = 0;								// hold the length of km4 data
+	uart_data[11] = (u32)g_uart1_buf;				// buffer to hold drained FIFO data
+
+	/* prepare buffers and notify KM4 to begin resume process */
+	DCache_Clean((u32)uart_data, sizeof(uart_data));
+	DCache_Clean((u32)g_uart1_buf, sizeof(g_uart1_buf));
+	ipc_send_message(IPC_AP_TO_NP, IPC_A2N_UART, &ipc_req_msg); 
+
+	/* wait for KM4 to finish the drain on its side */
+	while (1) {
+		DCache_Invalidate((u32)flag, sizeof(flag));
+		if (flag[0]) {
+			/* invalidate the cache to receive the drained FIFO data */
+			DCache_Invalidate((u32)g_uart1_buf, sizeof(g_uart1_buf));
+			DCache_Invalidate((u32)uart_data, sizeof(uart_data));
+
+			/* null terminate for safety */
+			g_uart1_buf[uart_data[10]] = 0;
+			break;
+		}
+	}
+
+	/* 
+	 * control has switched back from KM4 to CA32. 
+	 * KM4 has stopped reading the FIFO, so now we can drain it in CA32
+	 * no extra config on the peripheral should be done except detach attach irq as required
+	 */
+	u8 ch = 0;
+	g_uart1_dataleft = uart_data[10];
+	UART_TypeDef* uartx = UART_DEV_TABLE[uart_index_get(g_uart1priv.tx)].UARTx;
+
+	/* prevent PM transition while fetching remainder FIFO */
+	pm_timedsuspend(pm_domain_register("UART_PG_WAKE"), 5000);
```

Comment:
during uart bringup, a domain has been registered, how about refer to that pm id